hazelcast/src/main/java/com/hazelcast/nio/tcp/WriteHandler.java

```diff
@@ -172,7 +172,7 @@ private void unschedule() {
         // So the outputBuffer is empty, so we are going to unschedule ourselves.
         scheduled.set(false);
 
-        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty()) {
+        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty() && currentPacket == null) {
```

Comment:
Good points. 

These changes were made by me in HZ 3.4 at 12/16/2014. Reverting to the old write handlers could very well introduce new bugs since changes have been made in the IO system after I did the cleanup.

I propose that we give the IO system a complete cleanup.

The current code is complicated because of:
- the inherent nature of NIO
- accidental complexity caused by complex code.